content/software-testing-the-hardware-software-interface-part-2.md

```diff
@@ -0,0 +1,417 @@
+            echo 'Tests failed, please check the results.'
+        }
+
+        unstable {
+            junit 'telnet_test_results.xml'
+        }
+    }
+}
+```
+
+Which results in this nice CI job that is testing if our Verdin iMX8MP is properly booting! Cool, right?
+
+{{< figure src="/jenkins-result.png" title="Jenkins Pipeline Stages" alt="" width="100%" class="centered-figure">}}
+
+
+Of course you're free to use CircleCI, GitLab, GitHub whatever other service to do this. But the fact is, if you're
+doing Embedded at any scale, this architecture does scale very well.
+
+Having an automated service building and testing your hardware for every new `git push` (CI tools can automatically
+trigger jobs after new commits are added or even before during merge requests) allows you to safely push new code faster,
+and as I've hinted in the first article, it's more important than ever to safely create new patch releases due to
+cybersecurity regulations.
+
+I have to note that during these tests I was manually rebooting the board. Most places hook up the controller to relays
+or even other boards so that the controller itself resets and/or flashes the board at the beggining of a test. Notice
+that tests should start from *a known state*; if you're testing a software version 7.0.0, that version has to be installed
+when running tests.
+
+This pattern of "putting the board into a good state" is generally called Setup, and the process of cleaning up any
+mess your test left behind so as to not destroy the capability of a next test to put the board into a new known state is
+called Teardown.
+
+### Repeating the Pattern
+
+This example here was quite simple, but it would already be invaluable for any sized project. If you're a contractor,
+having tests like this is extremely helpful because you have a way to prove certain criteria were met during the project
+development - hard proof that your work was fully completed.
+
+Other more significant tests can simply repeat the pattern: want to test if the busses are working correctly? Hook the
+board up to a logic analyzer, the logic analyzer to the Controller and write some logic to test if your software is
+properly communicating with the hardware.
+
+A common case we'll talk about in the next installment are GUI validation tests, but as a spoiler you could, for example,
+setup a test where a screenshot is taken from within the board, moved over using `scp` to the runner job and use image
+comparison techniques to evaluate if the image is being rendered properly. Tests are mostly limited by the ingenuity of
+whoever is writing them.
+
+### Dual-Targetting with QEMU
+
+Dual-Targetting is a technique which involves isolating a given layer (generally the application layer) of a system to
+develop and test it further within a second target that is not generally the one actually being deployed.
+
+Why do this? Well, testing with hardware in the loop is quite time consuming. Developing evenmoreso. We want
```

Comment:
is evenmoreso a word, lol? Adding to my vocabulary, maybe....

Reply:
btw, I fixed it.